File: tests/test_audio_protocol.py
import struct

import pytest

from catchall.audio_protocol import (
    AUDIO_FRAME_TYPE,
    AUDIO_HEADER,
    AudioFrameError,
    decode_audio_frame,
)


def make_frame(samples: list[int], first_sample_index: int = 0) -> bytes:
    header = AUDIO_HEADER.pack(
        AUDIO_FRAME_TYPE,
        0,
        len(samples),
        first_sample_index,
    )
    pcm = struct.pack(f"<{len(samples)}h", *samples)
    return header + pcm

def test_decodes_pcm_audio_frame() -> None:
    frame = decode_audio_frame(
        make_frame([-32768, 0, 16384, 32767], first_sample_index=640)
    )

    assert frame.first_sample_index == 640
    assert frame.samples == pytest.approx([-1.0, 0.0, 0.5, 32767 / 32768])

def test_rejects_short_header() -> None:
    with pytest.raises(AudioFrameError, match="shorter than"):
        decode_audio_frame(b"\x01")

def test_rejects_unknown_frame_type() -> None:
    payload = bytearray(make_frame([0]))
    payload[0] = 99

    with pytest.raises(AudioFrameError, match="frame type"):
        decode_audio_frame(bytes(payload))

def test_rejects_incorrect_payload_length() -> None:
    payload = make_frame([1, 2, 3])

    with pytest.raises(AudioFrameError, match="length"):
        decode_audio_frame(payload[:-1])

File: tests/test_audio_protocol.js
import assert from "node:assert/strict";
import test from "node:test";

import {
    AUDIO_FRAME_TYPE,
    AUDIO_HEADER_BYTES,
    buildAudioFrame,
} from "../catchall/static/audio-protocol.js";


test("builds the binary audio fame header and payload", () => {
    const samples = Int16Array.from([
        -32768,
        -1,
        0,
        1,
        32767,
    ]);

    const frame = buildAudioFrame(samples, 640);
    const view = new DataView(frame);

    assert.equal(
        frame.byteLength,
        AUDIO_HEADER_BYTES + samples.length * 2
    );
    assert.equal(view.getUint8(0), AUDIO_FRAME_TYPE);
    assert.equal(view.getUint8(1), 0);
    assert.equal(view.getUint16(2, true), samples.length);
    assert.equal(view.getBigUint64(4, true), 640n);

    for (let index = 0; index < samples.length; index += 1) {
        assert.equal(
            view.getInt16(
                AUDIO_HEADER_BYTES + index * 2,
                true
            ),
            samples[index]
        );
    }
});

test("preserves sample indexes larger than 32 bits", () => {
    const firstSampleIndex = 2 ** 32 + 320;
    const frame = buildAudioFrame(
        Int16Array.from([0]),
        firstSampleIndex
    );
    const view = new DataView(frame);

    assert.equal(
        view.getBigUint64(4, true),
        BigInt(firstSampleIndex)
    );
});

test("rejects the wrong sample representation", () => {
    assert.throws(
        () => buildAudioFrame(new Float32Array([0]), 0),
        TypeError
    );
});

test("rejects empty audio frames", () => {
    assert.throws(
        () => buildAudioFrame(new Int16Array(0), 0),
        RangeError
    );
});

test("rejects invalid first sample indexes", () => {
    const samples = Int16Array.from([0]);

    assert.throws(
        () => buildAudioFrame(samples, -1),
        RangeError
    );
    assert.throws(
        () => buildAudioFrame(
            samples,
            Number.MAX_SAFE_INTEGER + 1
        ),
        RangeError
    );
});